tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc

```diff
@@ -54,6 +54,74 @@ __device__ __forceinline__ void AccumulateInto(
   CudaAtomicAdd(dest_scalar + 1, value.imag());
 }
 
+// SortedSegmentSumFunctor kernel reduces input data just as
+// UnsortedSegmentSumCustomKernel does except that input data
+// is partitioned along the outer reduction dimension. This is
+// because consecutive rows (elements in a row share the same
+// outer dimension index) in the flattened 2D input data likely
+// belong to the same segment in sorted segment sum operation.
+// Therefore such partitioning strategy has two advantages over
+// the UnsortedSegmentSumFunctor kernel:
+// 1. Each thread reduces across multiple rows before writing
+// answers to the global memory, we can therefore
+// write reduction results to global memory less often.
+// 2. We may know that the current thread is the only contributor
+// to an output element because of the increasing nature of segment
+// ids. In such cases, we do not need to use atomic operations
+// to write results to global memory.
+// In the flattened view of input data (with only outer and inner
+// dimension), every thread processes a strip of input data of
+// size OUTER_DIM_TILE_SIZE x 1. This strip runs across multiple
+// rows of input data and all reduction elements share one inner
+// dimension index.
+#define OUTER_DIM_TILE_SIZE 8
+#define CEIL_DIV(x, y) (1 + (((x)-1) / (y)))
```

Comment:
This produces the wrong answer when x == 0. The typical way is to write (x + y - 1) / y. 

In TF, the convention is to use Eigen::divup